Add OTLP to the specification (#794)
open-telemetry/opentelemetry-specification: specification/protocol/otlp.md

# OpenTelemetry Protocol Specification

> OpenTelemetry Protocol (OTLP) specification describes the encoding, transport

Reviewer: nit: oxford comma

```suggestion
OpenTelemetry Protocol (OTLP) specification describes the encoding, transport,
```

Author: Fixed.

and delivery mechanism of telemetry data between telemetry sources, intermediate

> nodes such as collectors and telemetry backends.

Reviewer: ```suggestion
nodes such as collectors, and telemetry backends.
```

Author: I may be wrong but I think it is not needed. From https://www.grammarly.com/blog/comma/
>Don’t separate two nouns that appear together as a compound subject or compound object.
>...you don’t need a comma when you’re simply listing two items.

## Table of Contents

> 

Reviewer: typycally generated TOcs has these comments in the beggining and end for proper updates. Or tags like this:

```
<details>
<summary>
Table of Contents
</summary>
```

What tool are you using?

Author: Markdown All in One extension for VS Code.

* [Table of Contents](#table-of-contents)
* [Protocol Details](#protocol-details)
  * [OTLP/gRPC](#otlpgrpc)
    * [Concurrent Requests](#concurrent-requests)
    * [Response](#response)
    * [Throttling](#throttling)
    * [gRPC Service and Protobuf Definitions](#grpc-service-and-protobuf-definitions)
    * [Default Port](#default-port)
  * [OTLP/HTTP](#otlphttp)
    * [Request](#request)
    * [Response](#response-1)
      * [Success](#success)
      * [Failures](#failures)
      * [Bad Data](#bad-data)
      * [Throttling](#throttling-1)
      * [All Other Responses](#all-other-responses)
    * [Connection](#connection)
    * [Concurrent Requests](#concurrent-requests-1)
    * [Default Port](#default-port-1)
* [Implementation Recommendations](#implementation-recommendations)
  * [Multi-Destination Exporting](#multi-destination-exporting)
* [Known Limitations](#known-limitations)
  * [Request Acknowledgements](#request-acknowledgements)
    * [Duplicate Data](#duplicate-data)
  * [Partial Success](#partial-success)
* [Future Versions and Interoperability](#future-versions-and-interoperability)
* [Glossary](#glossary)
* [References](#references)

> OTLP is a general-purpose telemetry data delivery protocol designed in the scope

Reviewer 2: Comparison with prior art is nice in proposals, or a rationale doc, but I guess we don't need it in a spec that's defining the protocol? If this was a more generic list of design goals (which may require copying in some of opencensus design goals which aren't changed) it would make sense though.

Author: I agree. Removed.

  rates change over time.

- Allows backpressure signalling from telemetry data destinations to sources.
  This is important for implementing reliable multi-hop telemetry data delivery
  all the way from the source to the destination via intermediate nodes, each
  having different processing capacity and thus requiring different data
  transfer rates.

- Defines both gRPC and HTTP 1.1 transports, allowing the clients to choose the
  best available transport.

## Protocol Details

OTLP defines the encoding of telemetry data and the protocol used to exchange
data between the client and the server.

This specification defines how OTLP is implemented over
[gRPC](https://grpc.io/) and HTTP 1.1 transports and specifies
[Protocol Buffers schema](https://developers.google.com/protocol-buffers/docs/overview)
that is used for the payloads.

OTLP is a request/response style protocols: the clients send requests, the
server replies with corresponding responses. This document defines one requests
and response type: `Export`.

### OTLP/gRPC

After establishing the underlying gRPC transport the client starts sending
telemetry data using unary requests using `Export*Request` messages

> ([`ExportTraceServiceRequest`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/trace/v1/trace_service.proto#L38)

Reviewer 3: I would avoid pining a specific commit to avoid outdated links. Maybe just refer to the proto repo in general?

Author: Good point. Added one link to the proto repo and removed the common-specific links.

of client/server nodes and aims to ensure that no data is lost in-transit
between the client and the server. Many telemetry collection systems have
intermediary nodes that the data must travel across until reaching the final
destination (e.g. application -> agent -> collector -> backend). End-to-end
delivery guarantees in such systems is outside of the scope of OTLP. The
acknowledgements described in this protocol happen between a single
client/server pair and do not span intermediary nodes in multi-hop delivery
paths._

#### Concurrent Requests

After sending the request the client MAY wait until the response is received
from the server. In that case there will be at most only one request in flight
that is not yet acknowledged by the server.

![Unary](img/otlp-sequential.png)

Sequential operation is recommended when simplicity of implementation is
desirable and when the client and the server are connected via very low-latency
network, such as for example when the client is an instrumented application and
the server is a OpenTelemetry Collector running as a local daemon (agent).

The implementations that need to achieve high throughput SHOULD support

> concurrent Unary calls to achieve higher throughput. The client SHOULD send new

Reviewer 2: As far as I know, for example in Java we don't ever send concurrent calls from the `BatchSpanProcessor`, but I believe that's our recommended implementation for high throughput scenarios too. Perhaps this is too much implementation detail for a spec? Just need to say implementations should support high throughput export.

Author: I think it is important to provide a bit more guidance and explain why. If one is not an expert in protocols and their implementations it is not immediately obvious that "high throughput" coupled with half-duplex protocol leads to a requirement to do concurrent requests.

for traces,
[`ExportMetricsServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/metrics/v1/metrics_service.proto#L44)
for metrics,
[`ExportLogsServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/ca6dcbbf390d8b3ce419a931eb351d189a3eb4fa/opentelemetry/proto/collector/logs/v1/logs_service.proto#L47)
for logs).

When an error is returned by the server it falls into 2 broad categories:
retryable and not-retryable:

- Retryable errors indicate that processing of telemetry data failed and the
  client SHOULD record the error and may retry exporting the same data. This can
  happen when the server is temporarily unable to process the data.

- Not-retryable errors indicate that processing of telemetry data failed and the
  client MUST NOT retry sending the same telemetry data. The telemetry data MUST
  be dropped. This can happen, for example, when the request contains bad data
  and cannot be deserialized or otherwise processed by the server. The client
  SHOULD maintain a counter of such dropped data.

> The server SHOULD indicate retryable errors using code

Reviewer: is it should here? There is no other way to indicate retryable error, correct? So perhaps this can be rephrased to 

```suggestion
When server chooses to indicate retryable error, it MUST be using code
```

Reviewer: this is not blocking as it's a matter of interpretation of this phrase.

Author: Changed to MUST.

[Unavailable](https://godoc.org/google.golang.org/grpc/codes) and MAY supply
additional
[details via status](https://godoc.org/google.golang.org/grpc/status#Status.WithDetails)
using
[RetryInfo](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L40)
containing 0 value of RetryDelay. Here is a sample Go code to illustrate:

```go
  // Do this on server side.
  st, err := status.New(codes.Unavailable, "Server is unavailable").
    WithDetails(&errdetails.RetryInfo{RetryDelay: &duration.Duration{Seconds: 0}})
  if err != nil {
    log.Fatal(err)
  }

  return st.Err()
```

> To indicate not-retryable errors the server is recommended to use code

Reviewer: Should we say that all other codes except one above are non retryable?

additional
[details via status](https://godoc.org/google.golang.org/grpc/status#Status.WithDetails)
using
[BadRequest](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L119).
Other gRPC status code may be used if it is more appropriate. Here is a sample
Go code to illustrate:

```go
  // Do this on server side.
  st, err := status.New(codes.InvalidArgument, "Invalid Argument").
    WithDetails(&errdetails.BadRequest{})
  if err != nil {
    log.Fatal(err)
  }

  return st.Err()
```

The server MAY use other gRPC codes to indicate retryable and not-retryable
errors if those other gRPC codes are more appropriate for a particular erroneous
situation. The client SHOULD interpret gRPC status codes as retryable or
not-retryable according to the following table:

|gRPC Code|Retryable?|
|---------|----------|
|CANCELLED|Yes|
|UNKNOWN|No|
|INVALID_ARGUMENT|No|
|DEADLINE_EXCEEDED|Yes|
|NOT_FOUND|No|
|ALREADY_EXISTS|No|
|PERMISSION_DENIED|No|
|UNAUTHENTICATED|No|
|RESOURCE_EXHAUSTED|Yes|
|FAILED_PRECONDITION|No|
|ABORTED|Yes|
|OUT_OF_RANGE|Yes|
|UNIMPLEMENTED|No|
|INTERNAL|No|
|UNAVAILABLE|Yes|
|DATA_LOSS|Yes|

> When retrying, the client SHOULD implement an exponential backoff strategy. An

Reviewer 2: Is this `MUST`? It seems weird that a client `MUST` respond to explicit throttling, but it's optional to use exponential backoff

Author: The client may decide to implement a different backoff strategy for whatever reason. SHOULD is typically used instead of MUST when deviating from the recommendation is undesirable but is not catastrophic. In this case it is not catastrophic to use a different backoff policy.

Author: Perhaps the same argument can be made about the response to throttling and it is also better to use SHOULD instead of MUST for that as well, but I think it is slightly more important to mandate that the client must honor server's desire to back off, and a bit less important to specify how exactly it should be honored.

Author: Either way, I think the difference in this particular case is small, happy to change if you have a strong opinion about it.

Author: Sounds good, resolving this comment.

exception to this is the Throttling case explained below, which provides
explicit instructions about retrying interval.

#### Throttling

OTLP allows backpressure signalling.

If the server is unable to keep up with the pace of data it receives from the
client then it SHOULD signal that fact to the client. The client MUST then
throttle itself to avoid overwhelming the server.

> To signal backpressure when using gRPC transport, the server SHOULD return an

Reviewer: again, I think MUST is better here

Author: Done.

error with code [Unavailable](https://godoc.org/google.golang.org/grpc/codes)
and MAY supply additional
[details via status](https://godoc.org/google.golang.org/grpc/status#Status.WithDetails)
using
[RetryInfo](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L40).
Here is a sample Go code to illustrate:

```go
  // Do this on server side.
  st, err := status.New(codes.Unavailable, "Server is unavailable").
    WithDetails(&errdetails.RetryInfo{RetryDelay: &duration.Duration{Seconds: 30}})
  if err != nil {
    log.Fatal(err)
  }

  return st.Err()
  
  ...

  // Do this on client side.
  st := status.Convert(err)
  for _, detail := range st.Details() {
    switch t := detail.(type) {
    case *errdetails.RetryInfo:
      if t.RetryDelay.Seconds > 0 || t.RetryDelay.Nanos > 0 {
        // Wait before retrying.
      }
    }
  }
```

When the client receives this signal it SHOULD follow the recommendations

> outlined in documentation for `RetryInfo`:

Reviewer: Please include an url link as well

Author: Done.

```
// Describes when the clients can retry a failed request. Clients could ignore
// the recommendation here or retry when this information is missing from error
// responses.
//
// It's always recommended that clients should use exponential backoff when
// retrying.
//
// Clients should wait until `retry_delay` amount of time has passed since
// receiving the error response before retrying.  If retrying requests also
// fail, clients should use an exponential backoff scheme to gradually increase
// the delay between retries based on `retry_delay`, until either a maximum
// number of retires have been reached or a maximum retry delay cap has been
// reached.
```

The value of `retry_delay` is determined by the server and is implementation
dependant. The server SHOULD choose a `retry_delay` value that is big enough to
give the server time to recover, yet is not too big to cause the client to drop
data while it is throttled.

#### gRPC Service and Protobuf Definitions

> gRPC service definitions [are here](https://github.com/open-telemetry/opentelemetry-proto/tree/master/opentelemetry/proto/collector).

Reviewer: Please add a phrase:

Please make sure to check the proto version and maturity level. Schemas for different signals may be at different maturity level - some stable, some in beta.

Author: Added.

request body is a Protobuf-encoded
[`ExportLogsServiceRequest`](https://github.com/open-telemetry/opentelemetry-proto/blob/ca6dcbbf390d8b3ce419a931eb351d189a3eb4fa/opentelemetry/proto/collector/logs/v1/logs_service.proto#L38)
message.

The client MUST set "Content-Type: application/x-protobuf" request header when
sending binary-encoded Protobuf or "Content-Type: application/json" request
header when sending JSON encoded Protobuf payload.

The client MAY gzip the content and in that case SHOULD include
"Content-Encoding: gzip" request header. The client MAY include
"Accept-Encoding: gzip" request header if it can receive gzip-encoded responses.

Non-default URL paths for requests MAY be configured on the client and server
sides.

JSON-encoded Protobuf payloads use proto3 standard defined
[JSON Mapping](https://developers.google.com/protocol-buffers/docs/proto3#json)
for mapping between Protobuf and JSON. `trace_id` and `span_id` is base64
encoded, not hex.

#### Response

Response body MUST be the appropriate serialized Protobuf message (see below for
the specific message to use in the Success and Failure cases).

> The server MUST set "Content-Type: application/x-protobuf" header if the

Reviewer 2: The server MUST send the same content type as it received?

Author: Good point. Added.

response body is binary-encoded Protobuf payload. The server MUST set
"Content-Type: application/json" if the response is JSON-encoded Protobuf
payload.

If the request header "Accept-Encoding: gzip" is present in the request the
server MAY gzip-encode the response and set "Content-Encoding: gzip" response
header.

##### Success

On success the server MUST respond with `HTTP 200 OK`. Response body MUST be
Protobuf-encoded
[`ExportTraceServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/trace/v1/trace_service.proto#L47)
message for traces, 
[`ExportMetricsServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/metrics/v1/metrics_service.proto#L44)
message for metrics and [`ExportLogsServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/ca6dcbbf390d8b3ce419a931eb351d189a3eb4fa/opentelemetry/proto/collector/logs/v1/logs_service.proto#L47)
message for logs.

> The server SHOULD respond with success no sooner than after successfully

Reviewer: Not blocking comment as this may be changed later if needed:

This effectively prevents from building udp-style http endpoints which may not be ideal. Especially for browsers when message needs to be accepted as fast as possible even if it will be dropped later. Browser or mobile app may not have another chance to send this message again.

Is it something that you think we can extend for later?

@open-telemetry/javascript-approvers what do you think here?

Author: >udp-style http endpoints

I am not familiar with the concept. Can you please clarify?

Reviewer: I mean when http endpoint just takes the payload and replies 200 OK right away. So browser is not waiting on deserialization.

Author: I understand. If there are good reasons to do that then it is indeed possible to do that. This clause is marked "SHOULD" which allows the implementor to deviate if they have good reasons to. Do you think "SHOULD" is too strong here?
(For the sake of making a progress we can probably merge as is, since this PR is merely bringing what already was approves as an OTEP to this repository and I can submit an issue to look into this later).

decoding and validating the request.

##### Failures

If the processing of the request fails the server MUST respond with appropriate
`HTTP 4xx` or `HTTP 5xx` status code. See sections below for more details about
specific failure cases and HTTP status codes that should be used.

Response body for all `HTTP 4xx` and `HTTP 5xx` responses MUST be a

> Protobuf-encoded

Reviewer 2: Maybe consider returning as response headers, similar to gRPC. Returning a binary payload for an error response is the easiest way to lose SRE friends :)

Author: Do you suggest to return the exact same data encoded in a more human-readable format in the response headers (i.e. Protobuf-JSON)? Or you suggest to include a human-readable text in a response header _in addition_ to the response body?

We could include a copy of `Status.message` (which is supposed to be human-readable) in a response header, but I think we still need to return the full `Status` (which may contain `details`) in the body.

Reviewer 2: Not going to block on this but still strongly recommend returning text somewhere for errors, headers or payload or otherwise. Protobuf already makes SREs hate gRPC in my experience, if we remove error headers too they'll really be unhappy.

Author: Added an issue to address later: https://github.com/open-telemetry/opentelemetry-specification/issues/808

[Status](https://godoc.org/google.golang.org/genproto/googleapis/rpc/status#Status)
message that describes the problem.

This specification does not use `Status.code` field and the server MAY omit
`Status.code` field. The clients are not expected to alter their behavior based
on `Status.code` field but MAY record it for troubleshooting purposes.

The `Status.message` field SHOULD contain a developer-facing error message as
defined in `Status` message schema.

The server MAY include `Status.details` field with additional details. Read

> below about what this field can contain in each specific failure case.

Reviewer 3: Not trying to block this, but why using the `Status` if the code is ignored. We should probably reconsider this section, an issue is good for the moment.

Author: Two reasons:
1. To keep the response more uniform with OTLP/gRPC.
2. To allow to include `Status.details`. This can contain important information for BadRequest response.

between retries must have a random jitter.

The client SHOULD keep the connection alive between requests.

Server implementations SHOULD accept OTLP/HTTP with binary-encoded Protobuf
payload and OTLP/HTTP with JSON-encoded Protobuf payload requests on the same
port and multiplex the requests to the corresponding payload decoder based on
the "Content-Type" request header.

Server implementations MAY accept OTLP/gRPC and OTLP/HTTP requests on the same
port and multiplex the connections to the corresponding transport handler based
on the "Content-Type" request header.

#### Concurrent Requests

To achieve higher total throughput the client MAY send requests using several
parallel HTTP connections. In that case the maximum number of parallel
connections SHOULD be configurable.

#### Default Port

The default network port for OTLP/HTTP is 55681. There is currently an [open
issue](https://github.com/open-telemetry/opentelemetry-collector/issues/1256) to
use the same port for OTLP/gRPC and OTLP/HTTP. In that case this spec will be
updated to use the same default port for OTLP/gRPC and OTLP/HTTP.

## Implementation Recommendations

> ### Multi-Destination Exporting

Reviewer 2: Is this related to OTLP?

Author: Strictly speaking no, it is applicable to any exporting protocol, but still important implementation detail when exporting to multiple destinations. If there is a better place for this information I can move it.

Reviewer 2: I think it could make sense closer to `Span Processor` or `Exporter` in this file https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md

Author: It is applicable not only to SDK but any exporting implementation, for example to the Collector, so I am not sure that would be the right place. I can delete from here for now if it looks misplaced.

Reviewer 2: To be honest, it's always hard to talk about "exporter" to team members since they get confused between SDK and collector. I consider them the same, either you export directly or you proxy through the collector. So indeed, SDK does seem like a bad file for this and other aspects of export, in another PR we should probably create a separate section for export specifically, as it stands for both SDK and collector

received yet. The client will typically choose to re-send such data to guarantee
delivery, which may result in duplicate data on the server side. This is a
deliberate choice and is considered to be the right tradeoff for telemetry data.

### Partial Success

The protocol does not attempt to communicate partial reception success from the
server to the client (i.e. when part of the data can be received by the server
and part of it cannot). Attempting to do so would complicate the protocol and
implementations significantly and is left out as a possible future area of work.

## Future Versions and Interoperability

OTLP will evolve and change over time. Future versions of OTLP must be designed
and implemented in a way that ensures that clients and servers that implement
different versions of OTLP can interoperate and exchange telemetry data. Old
clients must be able to talk to new servers and vice versa. If new versions of
OTLP introduce new functionality that cannot be understood and supported by
nodes implementing the old versions of OTLP the protocol must regress to the
lowest common denominator from functional perspective.

> When possible the interoperability SHOULD be ensured between all versions of

Reviewer 2: I'd say MUST, backwards compatibility is really important. If a completely unavoidable issue happens (e.g., security hole), I'd generally expect all the old versions to be obsolete anyways.

Author: Done.

   these fields. In many cases careful design of such schema changes and correct
   choice of default values for new fields is enough to ensure interoperability
   of different versions without nodes explicitly detecting that their peer node
   has different capabilities.

3. More significant changes must be explicitly defined as new optional
   capabilities in future OTEPs. Such capabilities SHOULD be discovered by
   client and server implementations after establishing the underlying
   transport. The exact discovery mechanism SHOULD be described in future OTEPs
   which define the new capabilities and typically can be implemented by making
   a discovery request/response message exchange from the client to server. The
   mandatory capabilities defined by this specification are implied and do not
   require a discovery. The implementation which supports a new, optional
   capability MUST adjust its behavior to match the expectation of a peer that
   does not support a particular capability.

The current version of OTLP is the initial version that describes mandatory
capabilities only. Implementations of this specification SHOULD NOT attempt to

> detect the capabilities of their peers and should operate as defined in this

Reviewer 2: I'm not sure where to look for "as defined in this document" since we don't talk about business logic. Is this paragraph needed?

Author: >I'm not sure where to look for "as defined in this document" 

"this document" is `otlp.md`, the document that one is reading. I can try to rephrase if it is not clear.

>since we don't talk about business logic.

I am confused by the reference to "business logic" and I don't understand what the problem is. 

>Is this paragraph needed?

Yes, I think it is important. In other words it says: "don't try to be clever and implement only parts of this spec and try to detect which parts of the spec your peers implemented. This is the first version, implement it fully as described".

Reviewer 2: I think the capabilities mentioned here are for example, the capability to send a `Status`. So it seems like this paragraph is talking more about business logic, how it handles `Status`, rather than the protocol itself. Another way to put it, I think even if an implementation did attempt to detect capabilities of their peers, I don't think that would conflict with anything else defined in this document. So I found it confusing what this is referring to. Maybe not `defined in this document` but `defined in this proto` (link to proto)?

Author: I re-read the paragraph and I think you are right, it does not add value to this document. Deleted.
